spinalcordtoolbox/moco/dl/inference.py

```diff
@@ -0,0 +1,180 @@
+        Input file:            {fname_data}
+        Output folder:         {os.path.abspath(ofolder)}
+        Mode:                  {mode}
+        Mask:                  {fname_mask if fname_mask else 'None'}
+        Reference image:       {fname_ref if fname_ref else 'first volume of input (t=0)'}
+        bvecs (dmri only):     {fname_bvecs if fname_bvecs else 'None'}
+        bvals (dmri only):     {fname_bvals if fname_bvals else 'None'}
+    """))
+
+    # Load NIfTI data
+    def load_nifti(p):
+        img = nib.load(p)
+        return img, img.get_fdata().astype(np.float32)
+
+    # Check required inputs
+    mov_img, mov_np = load_nifti(fname_data)
+    affine, header = mov_img.affine, mov_img.header
+
+    if fname_mask:
+        mask_img, mask_np = load_nifti(fname_mask)
+    else:
+        raise ValueError("[moco-dl] Spinal cord mask is required.")
+
+    if fname_ref:
+        fix_img, fix_np = load_nifti(fname_ref)
+    else:
+        # print("[moco-dl] No reference provided — will use first volume (t=0) of moving data.")
+        fix_np = mov_np[..., 0]
+        fix_np = np.repeat(fix_np[..., None], mov_np.shape[-1], axis=-1)
+
+    # Load DL model
+    print(f"[moco-dl] Loading checkpoint: {ckpt_path}")
+    model = DenseRigidReg()
+    state = torch.load(ckpt_path, map_location=device)
+    model.load_state_dict(state["state_dict"])
+    model = model.to(device)
+    model.eval()
+    model.warp = RigidWarp(mode="bilinear")
+
+    moving = torch.from_numpy(mov_np).unsqueeze(0).unsqueeze(0).to(device)
+    fixed = torch.from_numpy(fix_np).unsqueeze(0).unsqueeze(0).to(device)
+    mask = torch.from_numpy(mask_np).unsqueeze(0).unsqueeze(0).to(device)
+
+    # Run inference
+    print("[moco-dl] Starting inference...")
+    for _ in tqdm(range(1), desc="Running model"):
+        with torch.no_grad():
+            warped, flow, Tx, Ty = model(moving, fixed, mask)
+
+    warped = warped.squeeze().cpu().numpy()  # (H,W,D,T)
+    flow = flow.squeeze().cpu().numpy()
+    Tx = Tx.squeeze().cpu().numpy()  # (D,T)
+    Ty = Ty.squeeze().cpu().numpy()  # (D,T)
+
+    H, W, D, T = warped.shape
+    # Restore high-frequency detail from raw data
+    sharpened = np.copy(warped)
+    for t in range(T):
+        for d in range(D):
+            img_warped = warped[..., d, t]
+            img_raw = mov_np[..., d, t]
+            mask_slice = mask_np[..., d]
+
+            if np.count_nonzero(mask_slice) == 0:
+                sharpened[..., d, t] = img_warped
+                continue
+
+            raw_smooth = smooth(img_raw, sigmas=[0.5, 0.5])
+            texture = img_raw - raw_smooth
+            out = img_warped + 1.2 * texture
+            lo, hi = np.percentile(img_raw[mask_slice > 0], [0.5, 99.5])
+            out = np.clip(out, lo, hi)
+            sharpened[..., d, t] = out
+
+    matched = np.zeros_like(sharpened)
+    for t in range(T):
+        matched[..., t] = ski_exposure.match_histograms(sharpened[..., t], mov_np[..., t])
```

Comment:
This step is really interesting! It's been a while since I've seen "traditional" computer vision techniques applied to the output of a deep learning model.

Have you found that the output is improved with these extra steps? I'm curious what the output looks like before/after... :eyes: